### PR TITLE
Various improvements related to hibernation

### PR DESCRIPTION
**Pull request by keszybz**

### CODE REVIEW

File: src/shared/sleep-config.c [hunk 1]
```diff
@@ -274,24 +274,28 @@ static bool enough_memory_for_hibernation(void) {
...
 
-        /* zero this out in case we run on a file with no extents */
-        fiemap = new0(struct fiemap, 1);
+        /* struct fiemap is followed by struct fiemap_extent array.
+         * We use sizeof(struct fiemap_extent) to calculate allocation sizes,
+         * but put struct fiemap as the first entry. */
+        assert_cc(sizeof(struct fiemap) <= sizeof(struct fiemap_extent));
+
+        /* Zero this out in case we run on a file with no extents */
+        fiemap = calloc(1, sizeof(struct fiemap_extent));
```
Comment (poettering) [inline src/shared/sleep-config.c]: malloc0()?

Comment (poettering) [reply src/shared/sleep-config.c]: that said, maybe it'd be nicer not to hardcode the assert. Maybe instead:

```c
n_extra = DIV_ROUND_UP(sizeof(fiemap), sizeof(fiemap_extent));
fiemap = calloc(sizeof(fiemap_extent), n_extra);
```

This is not only formally nicer I think, but also more descriptive to what's going on here I think.

And then, `n_extra` can be used for the other, later calls too...

File: src/shared/sleep-config.c [hunk 2]
```diff
@@ -274,24 +274,28 @@ static bool enough_memory_for_hibernation(void) {
...
         if (!fiemap)
                 return -ENOMEM;
 
-        result_fiemap = new(struct fiemap, 1);
+        result_fiemap = malloc(sizeof(struct fiemap_extent));
```
Comment (poettering) [inline src/shared/sleep-config.c]: as with the above:

```c
result_fiemap = malloc_multiply(sizeof(struct fiemap_extent), n_extra);
```

File: src/shared/sleep-config.c (hunk 1 above)
Comment (keszybz) [reply src/shared/sleep-config.c]: I like this, it makes the whole thing shorter and more legible.

**Comment by keszybz** [comment]
```
-------
215/259 parse-hwdb                              FAIL     2.36 s
--- command ---
/builddir/build/BUILD/systemd-239/hwdb/parse_hwdb.py
--- stdout ---
/builddir/build/BUILD/systemd-239/hwdb/70-mouse.hwdb: 80 match groups, 109 matches, 93 properties
/builddir/build/BUILD/systemd-239/hwdb/60-keyboard.hwdb: 141 match groups, 216 matches, 718 properties
/builddir/build/BUILD/systemd-239/hwdb/70-touchpad.hwdb: 3 match groups, 5 matches, 3 properties
/builddir/build/BUILD/systemd-239/hwdb/70-joystick.hwdb: 2 match groups, 3 matches, 2 properties
/builddir/build/BUILD/systemd-239/hwdb/70-pointingstick.hwdb: 8 match groups, 22 matches, 11 properties
/builddir/build/BUILD/systemd-239/hwdb/60-sensor.hwdb: 49 match groups, 52 matches, 49 properties
/builddir/build/BUILD/systemd-239/hwdb/60-input-id.hwdb: 1 match groups, 1 matches, 2 properties
Cannot parse /builddir/build/BUILD/systemd-239/hwdb/60-evdev.hwdb: Expected stringEnd (at char 12880), (line:464, col:1)
/builddir/build/BUILD/systemd-239/hwdb/60-evdev.hwdb: 0 match groups, 0 matches, 0 properties
```
Fedora CI is failing because #8702, unrelated.

**Comment by keszybz** [comment]
Updated with an improved error message to close another issue.

File: src/login/logind-dbus.c [hunk 3]
```diff
@@ -1835,11 +1835,14 @@ static int method_do_shutdown_or_sleep(
 
         if (sleep_verb) {
                 r = can_sleep(sleep_verb);
+                if (r == -ENOSPC)
+                        return sd_bus_error_set(error, BUS_ERROR_SLEEP_VERB_NOT_SUPPORTED,
```
Comment (poettering) [inline src/login/logind-dbus.c:1839]: i wonder if this should get a new BUS_ERROR_xyz code?

Comment (keszybz) [reply src/login/logind-dbus.c:1839]: Dunno. If you think it should, I can add it, but I didn't see a reason atm.

**Pull request merged**